src/dd/Simulation.cpp

```diff
@@ -212,34 +246,25 @@ std::map<std::string, std::size_t> sample(const qc::QuantumComputation& qc,
 
 VectorDD simulate(const qc::QuantumComputation& qc, const VectorDD& in,
                   Package& dd) {
-  auto permutation = qc.initialLayout;
-  auto e = in;
+  qc::Permutation permutation = qc.initialLayout;
+  dd::VectorDD out = in;
   for (const auto& op : qc) {
-    // SWAP gates can be executed virtually by changing the permutation
-    if (op->getType() == qc::SWAP && !op->isControlled()) {
-      const auto& targets = op->getTargets();
-      std::swap(permutation.at(targets[0U]), permutation.at(targets[1U]));
-      continue;
+    if (isVirtuallySwappable(*op)) {
+      virtualSwap(*op, permutation);
+    } else {
+      out = applyUnitaryOperation(*op, out, dd, permutation);
     }
-
-    e = applyUnitaryOperation(*op, e, dd, permutation);
   }
-  changePermutation(e, permutation, qc.outputPermutation, dd);
-  e = dd.reduceGarbage(e, qc.getGarbage());
+
+  changePermutation(out, permutation, qc.outputPermutation, dd);
+  out = dd.reduceGarbage(out, qc.getGarbage());
 
   // properly account for the global phase of the circuit
-  if (std::abs(qc.getGlobalPhase()) > 0) {
-    // create a temporary copy for reference counting
-    auto oldW = e.w;
-    // adjust for global phase
-    const auto globalPhase = ComplexValue{std::polar(1.0, qc.getGlobalPhase())};
-    e.w = dd.cn.lookup(e.w * globalPhase);
-    // adjust reference count
-    dd.cn.incRef(e.w);
-    dd.cn.decRef(oldW);
+  if (fp phase = qc.getGlobalPhase(); hasGlobalPhase(phase)) {
```

Comment:
```suggestion
  if (const fp phase = qc.getGlobalPhase(); hasGlobalPhase(phase)) {
```
But this is subject to change anyway based on the comments above 🙂